src/rust/lib/frp/src/io/js.rs

```diff
@@ -0,0 +1,177 @@
+        }
+        let event_type = event_type.into();
+        Self {callback,element,event_type,logger}
+    }
+}
+
+impl Listener<dyn KeyboardEventCallback> {
+    /// Creates a new key down event listener.
+    pub fn new_key_down<F>(logger:impl AnyLogger, f:F) -> Self
+    where F : KeyboardEventCallback {
+        let boxed   = Box::new(f);
+        let closure = Closure::<dyn KeyboardEventCallback>::wrap(boxed);
+        Self::new(logger,"keydown",closure)
+    }
+
+    /// Creates a new key up event listener.
+    pub fn new_key_up<F>(logger:impl AnyLogger, f:F) -> Self
+    where F : KeyboardEventCallback {
+        let boxed   = Box::new(f);
+        let closure = Closure::<dyn KeyboardEventCallback>::wrap(boxed);
+        Self::new(logger,"keyup",closure)
+    }
+}
+
+impl Listener<dyn BlurEventCallback> {
+    /// Creates a blur event listener.
+    pub fn new_blur<F>(logger:impl AnyLogger, f:F) -> Self
+    where F : BlurEventCallback {
+        let boxed   = Box::new(f);
+        let closure = Closure::<dyn BlurEventCallback>::wrap(boxed);
+        Self::new(logger,"blur",closure)
+    }
+}
+
+impl<Callback:?Sized> Drop for Listener<Callback> {
+    fn drop(&mut self) {
+        let callback = self.callback.as_ref().unchecked_ref();
+        if self.element.remove_event_listener_with_callback(&self.event_type, callback).is_err() {
+            self.logger.warning("Couldn't remove event listener.");
+        }
+    }
+}
+
+
+// === Event Listener Options ===
+
+/// Retrun options for addEventListener function. See also
+/// https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener
+fn event_listener_options() -> web_sys::AddEventListenerOptions {
+    let mut options = web_sys::AddEventListenerOptions::new();
+    // We listen for events in capture phase, so we can decide ourself if it should be passed
+    // further.
+    options.capture(true);
+    // We want to prevent default action on wheel events, thus listener cannot be passive.
+    options.passive(false);
+    options
+}
+
+
+
+// ======================
+// === JsEventHandler ===
+// ======================
+
+/// Handler of currently processed js event.
+///
+/// Managed js event by this structure will be NOT propagated further to DOM elements, unless the
+/// `pass_to_dom` event will be emitted.
+///
+/// To make this class manage js event, you should wrap the closure passed as event listener using
+/// `make_event_handler` function.
+#[derive(Clone,CloneRef,Debug)]
+pub struct CurrentJsEvent {
+    /// Currently handled js event.
+    pub event       : frp::Stream<Option<web_sys::Event>>,
+    /// Emitting this signal while handling js event (`current_js_event` is Some) makes this event
+    /// pass to the DOM elements. Otherwise the js event propagation will be stopped.
+    pub pass_to_dom : frp::Source,
+    event_source    : frp::Source<Option<web_sys::Event>>,
+    network         : frp::Network,
+}
```

Comment:
why not use here our `define_endpoints` macro?

Reply:
Because it is in wrong crate so far.